rmgpy/data/kinetics/family.py

```diff
@@ -3478,48 +3791,68 @@ def getTrainingSet(self, thermoDatabase=None, removeDegeneracy=False, estimateTh
                             mol = mol.merge(react.molecule[0])
                         else:
                             mol = deepcopy(react.molecule[0])
-                    
-                            
-                    if mol.isSubgraphIsomorphic(root,generateInitialMap=True): #try product structures
-                        products = rxns[i].products
+
+                    if mol.isSubgraphIsomorphic(root,generateInitialMap=True) or (not fixLabels and getLabelFixedMol(mol,rootLabels).isSubgraphIsomorphic(root,generateInitialMap=True)): #try product structures
+                        products = [Species(molecule=[getLabelFixedMol(x.molecule[0],rootLabels)],thermo=x.thermo) for x in rxns[i].products]
                     else:
                         products = self.applyRecipe([s.molecule[0] for s in rxns[i].reactants],forward=True)
                         products = [Species(molecule=[p]) for p in products]
-                    
-                    prods = []
-                    for p in products: 
-                        for atm in p.molecule[0].atoms:
-                            if atm.label in rkeys:
-                                atm.label = reverseMap[atm.label]
-                    
-                        prods.append(Species(molecule=[p.molecule[0]]))
-                        
-                    rrev = Reaction(reactants=prods,products=rxns[i].reactants,kinetics=rxns[i].generateReverseRateCoefficient(),rank=rxns[i].rank)
-                    
+
+                    assert products != []
+                    prodmol = None
+                    for react in rxns[i].products:
+                        if prodmol:
+                            prodmol = prodmol.merge(react.molecule[0])
+                        else:
+                            prodmol = deepcopy(react.molecule[0])
+
+                    assert products != []
+                    if not prodmol.isSubgraphIsomorphic(root,generateInitialMap=True):
+                        mol = None
+                        for react in products:
+                            if mol:
+                                mol = mol.merge(react.molecule[0])
+                            else:
+                                mol = deepcopy(react.molecule[0])
+                        assert products != []
+                        if not mol.isSubgraphIsomorphic(root,generateInitialMap=True):
+                            for p in products:
+                                for atm in p.molecule[0].atoms:
+                                    if atm.label in rkeys:
+                                        atm.label = reverseMap[atm.label]
+
+                    reacts = [Species(molecule=[getLabelFixedMol(x.molecule[0],rootLabels)],thermo=x.thermo) for x in rxns[i].reactants]
+                    assert products != []
```

Comment:
Is it necessary to assert that `products != []` four times? It doesn't seem like anything modifies `products` after it's defined.

Reply:
Sorry, those are leftover debug.  Removed